heartbeat/autodiscover/builder/hints/monitors.go

```diff
@@ -134,28 +141,46 @@ func (hb *heartbeatHints) getProcessors(hints mapstr.M) []mapstr.M {
 	return utils.GetConfigs(hints, "", "processors")
 }
 
-func (hb *heartbeatHints) getHostsWithPort(hints mapstr.M, port int) []string {
-	var result []string
+func (hb *heartbeatHints) getHostsWithPort(hints mapstr.M, port int) ([]string, error) {
 	thosts := utils.GetHintAsList(hints, "", hosts)
-	// Only pick hosts that have ${data.port} or the port on current event. This will make
-	// sure that incorrect meta mapping doesn't happen
+	mType := utils.GetHintString(hints, "", scheme)
+
+	// We can't reliable detect duplicated monitors since we don't have all ports/hosts,
+	// relying on scheduler deduping monitors, see https://github.com/elastic/beats/pull/29041
+	hostSet := map[string]struct{}{}
 	for _, h := range thosts {
-		if strings.Contains(h, "data.port") || strings.Contains(h, fmt.Sprintf(":%d", port)) ||
-			// Use the event that has no port config if there is a ${data.host}:9090 like input
-			(port == 0 && strings.Contains(h, "data.host")) {
-			result = append(result, h)
-		} else if port == 0 && !strings.Contains(h, ":") {
-			// For ICMP like use cases allow only host to be passed if there is no port
-			result = append(result, h)
-		} else {
-			hb.logger.Warn("unable to frame a host from input host: %s", h)
+		if mType == "icmp" && strings.Contains(h, ":") {
+			hb.logger.Warnf("ICMP scheme does not support port specification: %s", h)
+			continue
+		} else if strings.Contains(h, ":") && strings.Contains(h, "data.port") && port == 0 {
+			// 0 is not -technically- a user-defined port, skip if a dynamic port is set
+			continue
+		} else if strings.Contains(h, ":") && !strings.Contains(h, "data.port") {
```

Comment:
according to https://www.elastic.co/guide/en/beats/heartbeat/current/configuration-autodiscover-hints.html we should check for `${data.port}` and `${data.host}` not just the string `data.port` or `data.host`